drivers: wifi: esp32: Add WiFi event information for station connect and disconnect

chirambaht opened the pull request:
The ESP32's WiFi event information system has now been linked with the Zephyr WiFi events system for station related operations. This allows users of the ESP32 wifi driver to obtain more precise WiFi connection related event information such as incorrect passwords, timeouts and more.

The work presented here was guided by the ESP32's WiFi  event documentation provided here: https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/wifi.html#event-handling. 

In order to validate this, I ran the wifi_shell example with the appropriate logging levels for WiFi and passed in various different network credentials to see if the appropriate events were raised with the correct information.

chirambaht commented:
@rodrigopex

jukkar commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -662,10 +726,10 @@ static int esp32_wifi_ap_enable(const struct device *dev,
 	wifi_config_t wifi_config = {
 		.ap = {
-			.max_connection = 5,
+				.max_connection = 5,
```
Unrelated change, and it looks like the indentation is now wrong

jukkar commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -662,10 +726,10 @@ static int esp32_wifi_ap_enable(const struct device *dev,
 			.channel = params->channel == WIFI_CHANNEL_ANY ?
 				0 : params->channel,
-		},
+			},
```
ditto

jukkar commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -981,7 +1045,7 @@ NET_DEVICE_DT_INST_DEFINE(0,
 		&esp32_data, NULL, CONFIG_WIFI_INIT_PRIORITY,
 		&esp32_api, ETHERNET_L2,
-		NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+			  NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
```
ditto

jukkar commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -965,7 +1029,7 @@ static const struct wifi_mgmt_ops esp32_wifi_mgmt = {
 
 static const struct net_wifi_mgmt_offload esp32_api = {
-	.wifi_iface.iface_api.init	  = esp32_wifi_init,
+	.wifi_iface.iface_api.init = esp32_wifi_init,
```
Unrelated change here too

sylvioalves commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -307,20 +307,84 @@ static void esp_wifi_handle_sta_connect_event(void *event_data)
 	}
 
-	LOG_DBG("Disconnect reason: %d", event->reason);
+	wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface, WIFI_REASON_DISCONN_SUCCESS);
```
Perhaps you could return the disconnect event value from both calls you created and have this single call here with that value.
In current approach:
1) `wifi_mgmt_raise_disconnect_result_event` is being called 2x.
2) You have multiple calls to `wifi_mgmt_raise_disconnect_result_event` when you could use a single one.

Would you mind analyzing if the above suggestion is acceptable?

chirambaht committed:
drivers: wifi: esp32: Add WiFi events for station connect and disconnect

Added events for when the station iface is used to connect and disconnect.
These events will feed the NET_EVENT_WIFI_CONNECT_RESULT and
NET_EVENT_WIFI_DISCONNECT_RESULT events to send information
to applications making use of them

Signed-off-by: Humphrey Chiramba <humphrey@plentify.co.za>

sylvioalves commented:
@chirambaht, can you clarify what is the situation that does not allow merging both disconnect events? I guess each event error is dependent on the connection itself. From the changes and PR description, sounds we have the same event happening for both disconnect and not-accepted-connection. Which events are that?
I see the mutual events in there are `WIFI_REASON_ASSOC_LEAVE` and `WIFI_REASON_AUTH_EXPIRE` only.
Sorry to extending this.
